File: CS4700_ProgrammingLanguages/6-JuliaJulia/main.cpp
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <string>

#include "ppmToBmp.hpp" // provides the function int ppmToBmp(std::string ppmFileName);

using namespace std;

struct Color {
    int r;
    int g;
    int b;
};

struct MandelbrotConfig {
    //
    // These values come from the configuration file
    //
    int pixels;            // dimensions of square image, in pixels

    double midX, midY;     // center of image in imaginary coords
    double axisLen;        // width and height of image in imaginary units

    int maxIterations;     // max number of iterations to test for escape

    Color colorOne;        // color of pixels outside of Mandelbrot set
    Color colorTwo;        // color of pixels within Mandelbrot set

    string outputPPMfile;  // name of PPM file to create


    //
    // These values can be calculated from the above, and will
    // make your algorithm more convenient and efficient
    //
    double minX;
    double minY;                          // X coords in imaginary units of left & right sides of image
    double maxX;
    double maxY;                          // Y coords in imaginary units of top & bottom sides of image
    double pixelSize;                     // how many imaginary units of length/height each pixel takes
    double step_r;                        // how far the red channel changes between iterations
    double step_g;                        // how far the green channel changes between iterations
    double step_b;                        // how far the blue channel changes between iterations
    
};



// Read fractal config file and return a MandelbrotConfig struct
MandelbrotConfig readConfig(string fileName){
    MandelbrotConfig cfg;
        
    //Open File and read info from it
    ifstream fin(fileName);
    if(!fin) {
        cout << "File Not Found!" << endl;
        return cfg;
    }
    
    //read in given values
    fin >> cfg.pixels >> cfg.midX >> cfg.midY >> cfg.axisLen >> cfg.maxIterations >> cfg.colorOne.r >> cfg.colorOne.g >> cfg.colorOne.b 
    >> cfg.colorTwo.r >> cfg.colorTwo.g >> cfg.colorTwo.b >> cfg.outputPPMfile;
    
    //calculate values
    cfg.minX = cfg.midX - (cfg.axisLen/2);
    cfg.minY = cfg.midY - (cfg.axisLen/2);  
    cfg.maxX = cfg.midX + (cfg.axisLen/2);
    cfg.maxY = cfg.midY + (cfg.axisLen/2); 
    cfg.pixelSize = cfg.axisLen/cfg.pixels;
    cfg.step_r = static_cast <double>((cfg.colorTwo.r - cfg.colorOne.r))/cfg.maxIterations; 
    cfg.step_g = static_cast <double>((cfg.colorTwo.g - cfg.colorOne.g))/cfg.maxIterations;
    cfg.step_b = static_cast <double>((cfg.colorTwo.b - cfg.colorOne.b))/cfg.maxIterations;
    
    
    return cfg;
}

// Helper function: given a coordinate corresponding to an imaginary number,
// run the Escape Time algorithm and count how many iterations this number takes
int countIterations(MandelbrotConfig cfg, int i, int j){
    double xtemp = 0.0;
    double x = 0.0;
    double y = 0.0;
    int iteration = 0;
    // Scale current pixel coordinates to lie inside the complex Mandelbrot X/Y scale
    double x0 = cfg.minX + j * cfg.pixelSize;
    double y0 = cfg.maxY - i * cfg.pixelSize;
    
    //find out number of iterations
    while (((x*x + y*y) < 4) && (iteration < cfg.maxIterations)){
              xtemp = x*x - y*y + x0;
              y = 2*x*y + y0;
              x = xtemp;
              iteration = iteration + 1;
    }
    return iteration;
}

// Helper function: decide what color corresponds to a given number of iterations
Color getPixelColor(MandelbrotConfig cfg, int iterations){
    Color pixelColor;
    
    pixelColor.r = cfg.colorOne.r + (cfg.step_r*iterations);
    pixelColor.b = cfg.colorOne.b + (cfg.step_b*iterations);
    pixelColor.g = cfg.colorOne.g + (cfg.step_g*iterations);
    
    return pixelColor;
}

// Create the PPM file, including header, and build the image pixel by pixel
// Loop over the grid of pixels and call the above helper functions as needed
bool drawMandelbrot(MandelbrotConfig cfg){
    double x0;
    double y0;
    double xtemp;
    int y= 0;
    int iteration;
    Color pixelColor;
    
    //Open PPM file 
    ofstream fout (cfg.outputPPMfile);
    
    //Create Header
    fout << "P3" << endl << cfg.pixels << " " << cfg.pixels << endl << "255" << endl;
    
    //For each row in image, do:
    for (int i=0; i< cfg.pixels; i++){
        
        //For each col in row, do:
        for (int j=0; j<cfg.pixels; j++){
           
           iteration = countIterations(cfg, i, j);
    
           pixelColor = getPixelColor(cfg, iteration);
           
           //add pixel to file
           fout << setw(4) << left << static_cast<int> (pixelColor.r) << setw(4) << left << static_cast<int> (pixelColor.g) 
                << setw(6)<< left << static_cast<int> (pixelColor.b);
        }
    
    }
    fout.close();
    return true;
}

int main(int argc, char* argv[]) {
    string filename;

    //Read in config file location from use
    cout << "Please enter your file name:";                                                    
    cin >> filename;                                                                           

    // Create a configuration struct from the file
    MandelbrotConfig cfg = readConfig(filename);
    
    // Compute and write specified mandelbrot image to PPM file
    if (drawMandelbrot(cfg)) {
        // use the provided function to create a BMP image
        ppmToBmp(cfg.outputPPMfile);
    }
    else {
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

File: CS4700_ProgrammingLanguages/6-JuliaJulia/main.jl
include("rgb.jl") # includeing files is very similar to C++
include("hsv.jl")
include("config.jl")

function readConfig(fileName)
  f = open(fileName)
  pixles = split(readline(f)," ")
  cval = split(readline(f)," ")
  mid = split(readline(f)," ")
  axisLen = split(readline(f)," ")
  c1 = split(readline(f), " ")
  c2 = split(readline(f), " ")

  pixlesX = parse(Int, pixles[1])
  pixlesY = parse(Int, pixles[2])
  cvalR = parse(Float64, cval[1])
  cvalI = parse(Float64, cval[2])
  c = cvalR + (cvalI)im
  midX = parse(Float64, mid[1])
  midY = parse(Float64, mid[2])
  axisLenX = parse(Float64, axisLen[1])
  axisLenY = parse(Float64, axisLen[2])
  maxIters = parse(Int, split(readline(f), " ")[1])

  color1 = RGB(parse(Int, c1[1]), parse(Int, c1[2]), parse(Int, c1[3]))
  color2 = RGB(parse(Int, c2[1]), parse(Int, c2[2]), parse(Int, c2[3]))

  outputFile = split(readline(f), " ")[1]

  minX = midX - axisLenX / 2
  maxX = midX + axisLenX / 2
  minY = midY - axisLenY / 2
  maxY = midY + axisLenY / 2
  pxlSizeX = axisLenX / pixlesX
  pxlSizeY = axisLenY / pixlesY
  step_r = (color2.r - color1.r) / maxIters
  step_g = (color2.g - color1.g) / maxIters
  step_b = (color2.b - color1.b) / maxIters

  Config(pixlesX, pixlesY, c, midX, midY, axisLenX, axisLenY, maxIters, color1, color2, outputFile, minX, maxX, minY, maxY, pxlSizeX, pxlSizeY, step_r, step_g, step_b)

end

function hsvTOrgb(hsv::HSV)
  H = hsv.h
  S = hsv.s
  V = hsv.v

  (H == 360.0)?(H = 0.0):(H /= 60.0)
  fract = H - floor(H)

  P = floor(V*(1.0 - S)*255)
  Q = floor(V*(1.0 - S*fract)*255)
  T = floor(V*(1.0 - S*(1.0 - fract))*255)

  if (0 <= H && H < 1)
    return RGB(V, T, P)
  elseif (1 <= H && H < 2)
    return RGB(Q,V,P)
  elseif (2 <= H && H < 3)
    return RGB(P,V,T)
  elseif (3 <= H && H < 4)
    return RGB(P,Q,V)
  elseif (4 <= H && H < 5)
    return RGB(T,P,V)
  elseif (5 <= H && H < 6)
    return RGB(V,P,Q)
  else
    return RGB(0,0,0)
  end
end

function gradient1(cfg, iters)

    r = floor(cfg.color1.r + (cfg.step_r*iters))
    b = floor(cfg.color1.b + (cfg.step_b*iters))
    g = floor(cfg.color1.g + (cfg.step_g*iters))

    return RGB(r,g,b)
end

function getColor(cfg, iters, style)
  if style == 0
    return hsvTOrgb(HSV((iters % 256), 1, (1 * (iters < cfg.maxIters))));
  elseif style == 1
    return gradient1(cfg, iters)
  end
end

function getIters(cfg, row, col)
  zr = cfg.minX + (row - 1) * cfg.pxlSizeX
  zi = cfg.minY + (col - 1) * cfg.pxlSizeY
  z = zr + (zi)im # native support for complex numbers

  for n = 1:cfg.maxIters
    if abs2(z) > 4
      return n-1
    end
    z = z^2 + cfg.c # ^ operator does powers
  end
  return cfg.maxIters

end

function main()
  cfg = readConfig("config.txt")

  println(cfg.c)

  fout = open(cfg.outputFile,"w")
  write(fout, "P3\n$(cfg.pixlesX) $(cfg.pixlesY)\n255\n") # string interpolation is very convenient

  for col = 1:cfg.pixlesY
    for row = 1:cfg.pixlesX

    iters = getIters(cfg, row, col)
    color = getColor(cfg, iters, 0)
    write(fout, "$(color.r) $(color.g) $(color.b)\t")

    end
  end

  close(fout)
end

main()

# Julia does not do implicit type casting
# Having a REPL is amazing for learning a language

